### ISSUE TITLE
Firefox ESR 31 used despite firefox: "39.0" plugin in .travis.yml

### ISSUE THREAD

**Issue by miketaylr** [travis-ci/travis-ci#4649]
See for example, https://travis-ci.org/webcompat/webcompat.com/builds/75348340 (or https://s3.amazonaws.com/archive.travis-ci.org/jobs/75348341/log.txt):

```
[34m[1mInstalled Firefox version[0m
firefox 31.0esr
```

As well as 

```
[0K[33;1mInstalling Firefox v39.0[0m
```

https://github.com/webcompat/webcompat.com/blob/master/.travis.yml#L16 has:

``` yml
addons:
  firefox: "39.0"
```

But you can see that it's actually just using 31:

![](https://cldup.com/wUI5Ne-udn-3000x3000.png)

**Comment by BanzaiMan** [comment]
How are your tests invoking `firefox`? https://travis-ci.org/BanzaiMan/travis_production_test/builds/76097930#L114 shows that `/usr/local/bin/firefox` (which is at the top of `$PATH`) does indeed point to version 39.0. There is also `/usr/local/firefox-31.0esr/firefox`, but I am not sure how you might be invoking that instead.

**Comment by vladikoff** [comment]
@miketaylr as part of your script it might be a good idea to run `which firefox`, and here is how to force the binary path: https://github.com/mozilla/fxa-content-server/blob/master/tests/intern.js#L94

**Comment by miketaylr** [comment]
@BanzaiMan I'm just pointing to `firefox` from the `browserName` of the Intern.js `environments` array (https://theintern.github.io/intern/#option-environments):

https://github.com/webcompat/webcompat.com/blob/master/tests/intern.js#L30

@vladikoff do you know how Intern resolves the environment to a `$PATH`?

**Comment by evilaliv3** [comment]
i'm experiencing the same issue trying to use the latest and using protractor: https://api.travis-ci.org/jobs/129151209/log.txt?deansi=true

i'm now satisfied using this fix on the PATH:
https://github.com/globaleaks/GlobaLeaks/commit/1788becc5dddb04a3774af63d7c03abbaf84efc9

https://api.travis-ci.org/jobs/129176010/log.txt?deansi=true

**Comment by AutomatedTester** [comment]
We are now on Firefox 45ESR...

This is now creating an issue in https://github.com/mozilla/geckodriver/issues/85

**Comment by BanzaiMan** [comment]
@AutomatedTester I understand that we are behind in updating the build environment, including Firefox, but this issue has nothing to do with the issue you are seeing.

This issue is about, despite using the `firefox` addon, the version invoked is not what you specify.

Your issue is most likely fixed by using the addon in your [`.travis.yml`](https://github.com/mozilla/geckodriver/blob/c007252f0e3758a9b028097f4b3529b1946bf516/.travis.yml).

**Comment by bjdevil21** [comment]
Also having this problem. Using Selenium 2.53.0 on Travis-CI. Using .travis.yml addon: firefox: "46.0.1", I get Firefox v38. There was no mention of adding the newer FF to $PATH in the output (see truncated CI output below):

`Installing Firefox 46.0.1
1.47s$ wget -O /tmp/firefox-46.0.1.tar.bz2 $FIREFOX_SOURCE_URL
--2016-06-17 05:31:24--  https://download.mozilla.org/?product=firefox-46.0.1&lang=en-US&os=linux64 ... 2016-06-17 05:31:25 (44.9 MB/s) - `/tmp/firefox-46.0.1.tar.bz2' saved [52884219/52884219] ....
Driver info: org.openqa.selenium.firefox.FirefoxDriver ...
      Capabilities [{platform=LINUX, acceptSslCerts ... **version=38.4.0**}] (WebDriver\Exception\UnknownError)`

FF46 + Selenium 2.53.0 work fine locally when running Behat tests with the same Selenium + FF versions.

**Comment by BanzaiMan** [comment]
How does Selenium invoke Firefox?

**Comment by bjdevil21** [comment]
It was with:
```
addons:
  firefox: "46.0.1"
```
...in a .travis.yml, which isn't working. I Get the FF 38.4.0 version.

To get it to work with 46.0.1, I used the following workaround in .travis.yml:
`  - sudo mkdir /usr/local/bin/firefoxhack
  - sudo wget --directory-prefix=/tmp http://ftp.mozilla.org/pub/firefox/releases/46.0.1/linux-x86_64/en-US/firefox-46.0.1.tar.bz2
  - sudo tar -xjf /tmp/firefox-46.0.1.tar.bz2 -C /usr/local/bin/firefoxhack --strip-components=1
  - export PATH="/usr/local/bin/firefoxhack:$PATH"`

**Comment by shane-tomlinson** [comment]
We saw the same problem, we installed Firefox 46 using `addons`:

```yml
addons:
  firefox: "46.0"
```

In our `script` section, I printed the version of both `firefox` and `firefox-bin`:

```yml
script:
  - which firefox
  - firefox --version
  - which firefox-bin
  - firefox-bin --version
```

`firefox` version was `Mozilla Firefox 46.0`
`firefox-bin` version was `Mozilla Firefox 31.0`

When The Intern/Selenium fired up its instance of Firefox, I had it print it's ua string and saw `Mozilla/5.0 (X11; Linux x86_64; rv:31.0) Gecko/20100101 Firefox/31.0`

Here's the output run: https://travis-ci.org/mozilla/fxa-content-server/builds/167695971

We got around this problem by passing the result of `which firefox` to The Intern's [config.capabilities.firefox_binary` configuration option](https://github.com/mozilla/fxa-content-server/blob/master/tests/intern.js#L74) as mentioned by @vladikoff [above](https://github.com/travis-ci/travis-ci/issues/4649#issuecomment-132207090).

**Comment by BanzaiMan** [comment]
I think this is worth documenting. @shane-tomlinson Would you like to give us a hand in polishing up https://github.com/travis-ci/docs-travis-ci-com/blob/gh-pages/user/firefox.md?

**Comment by nenadalm** [comment]
another option to setting ```config.capabilities.firefox_binary``` would be to replace argument ```--binary``` for ```geckodriver``` by some wrapper https://github.com/nenadalm/Warehouse/blob/1378825e8f0c32fbd2878697753939fef1fd9420/.travis.yml#L34

**Comment by BanzaiMan** [comment]
I will try updating the addon to create symlink for `firefox-bin` as well as `firefox`. This might solve the issue.

**Comment by BanzaiMan** [comment]
Actually, there are more executables than just `firefox` and `firefox-bin` in the `firefox` directory of the archive, so I think it is better to just add the expanded archive directory to `$PATH`.

**Comment by henrahmagix** [comment]
For my uses with Protractor, I used:
```yaml
script:
    - alias firefox-bin='firefox'
    - protractor stuff
```

**Comment by nolanlawson** [comment]
FWIW we have been battling this problem in PouchDB as well (https://github.com/pouchdb/pouchdb/pull/5947). I've tried several different combinations, and I can't get our Selenium tests to consistently use the right version of Firefox, but thanks to comments in this thread I've been able to get it to work at least a majority of the time. Setting an explicit `firefox_binary` option in the Selenium config to point to `/usr/local/bin/firefox` seems to be the best strategy. (We're using the `selenium-standalone` npm package.)

It'd be great if Travis provided some way to install a later version of Firefox and _replace_ the existing Firefox version. Apparently having the two installed at the same time causes a lot of wonkiness, although I'm still not sure what the actual problem is.
